### PR TITLE
add binary search implemented by recursion in javascript

### PR DESCRIPTION
**Pull request by majorcool**
**Fixes issue:**

**Changes:**

### CODE REVIEW

File: code/search/src/binary_search/binary_search.js [hunk 1]
```diff
@@ -1,18 +1,49 @@
...
-        mid = Math.floor((lo + hi) / 2, 10);
-        element = array[mid];
-        if (element < key) {
-            lo = mid + 1;
-        } else if (element > key) {
-            hi = mid - 1;
-        } else {
-            return mid;
-        }
+  var lo = 0,
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: Use `let`, not `var`.

File: code/search/src/binary_search/binary_search.js [hunk 2]
```diff
@@ -1,18 +1,49 @@
...
+    hi = array.length - 1,
+    mid,
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: `mid` can be moved inside the loop. So can `element` be.

File: code/search/src/binary_search/binary_search.js [hunk 3]
```diff
@@ -1,18 +1,49 @@
...
+
+// implementation by recursion
+/**
+ *
+ * @param {* number[]} arr - the sorted array to be searched in
+ * @param {* number} value - the value to be searched
+ * @param {* number} low - the start index of the search range
+ * @param {* number} high - the end index of the search range
+ */
+function binarySearchByRecursion(arr, value, low, high) {
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: In ES6, Default parameters can be placed directly in the function. ie. `...Recursion(arr, value, low = 0, high=arr.length - 1)`. Also don't call the function `binarySearchByRecursion`, just call it `binarySearchRecursion`.

File: code/search/src/binary_search/binary_search.js [hunk 4]
```diff
@@ -1,18 +1,49 @@
...
+  var start = low === undefined ? 0 : low;
+  var end = high === undefined ? arr.length - 1 : high;
+  var middle = Math.floor((start + end) / 2);
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: Sometimes you use `mid` for a name, sometimes `middle`. Pick one.

Comment (majorcool) [reply code/search/src/binary_search/binary_search.js]: OK,thx! I'll fix it ASAP.

**Comment by arnavb** [comment]
@majorcool Why did you close this PR?

**Comment by majorcool** [comment]
@arnavb Sorry if I did some misoperation. 'Cause I don't quite know what the right process is. Anyway, I created a new PR in which I fixed the problems you commented above.

**Pull request closed without merging**